### PR TITLE
Removing an attachment is an impure operation

### PR DESCRIPTION
**Pull request by dsainati1**
This was not properly marked as impure in the checker. 

Thanks to @oebeling for the report. 

______

- [ ] Targeted PR against `master` branch
- [ ] Linked to Github issue with discussion and accepted design OR link to spec that describes this work
- [ ] Code follows the [standards mentioned here](https://github.com/onflow/cadence/blob/master/CONTRIBUTING.md#styleguides)
- [ ] Updated relevant documentation 
- [ ] Re-reviewed `Files changed` in the Github PR explorer
- [ ] Added appropriate labels

### CODE REVIEW

File: runtime/sema/check_remove_statement.go [hunk 1]
```diff
@@ -31,6 +31,8 @@ func (checker *Checker) VisitRemoveStatement(statement *ast.RemoveStatement) (_
 		})
 	}
 
+	checker.ObserveImpureOperation(statement)
```
Comment (turbolent) [inline runtime/sema/check_remove_statement.go:34]: Maybe it's worth verifying checking of all statements and expressions to see if there are similar cases in the code base.

**Review by turbolent — APPROVED** [review]
👍

File: runtime/tests/checker/attachments_test.go [hunk 2]
```diff
@@ -4578,3 +4578,94 @@ func TestCheckAttachmentRemoveLossTracking(t *testing.T) {
...
+			`
+			struct S {}
+			attachment Test for S {}
+			view fun foo(s: S) {
+				remove Test from s
+			}`,
+		)
+
+		errs := RequireCheckerErrors(t, err, 1)
+		assert.IsType(t, &sema.PurityError{}, errs[0])
```
Comment (SupunS) [inline runtime/tests/checker/attachments_test.go:4669]: Why does this need report a purity error?

Comment (dsainati1) [reply runtime/tests/checker/attachments_test.go:4669]: I suppose it doesn't, since this `s` is a copy. Will update.

**Review by SupunS — APPROVED** [review]

File: runtime/tests/checker/attachments_test.go [hunk 3]
```diff
@@ -4665,6 +4665,65 @@ func TestCheckAttachmentPurity(t *testing.T) {
...
+
+		t.Parallel()
+
+		_, err := ParseAndCheck(t,
+			`
+			struct S {}
+			attachment Test for S {}
+			var s: S = S()
+			view fun foo() {
+				remove Test from s
```
Comment (turbolent) [inline runtime/tests/checker/attachments_test.go:4681]: Are the semantics for `remove` on a value-type (e.g. struct) that the variable is re-assigned? Is `s` after the removal a new copy of `s` without the attachment (and the original value is unmodified, but "lost")?

Comment (dsainati1) [reply runtime/tests/checker/attachments_test.go:4681]: The variable is mutated by the `remove`, so a globally scoped or field-scoped struct will be modified.

**Comment by dsainati1** [comment]
cc @turbolent does this explanation make sense?

**Review by turbolent — APPROVED** [review]
Yep! Nice 👌

**Pull request merged**